packages/core/src/config.ts

```diff
@@ -955,64 +959,113 @@ const composeEntryConfig = async (
+  const jsRedirectPath = redirect.js?.path ?? true;
+  const jsRedirectExtension = redirect.js?.extension ?? true;
 
-  const isStyleRedirect = redirect.style ?? true;
+  type Resolver = GetAsyncFunctionFromUnion<
+    ReturnType<NonNullable<Rspack.ExternalItemFunctionData['getResolve']>>
+  >;
+  let resolver: Resolver | undefined;
 
   return {
-    output: {
-      externals: [
-        (data: any, callback: any) => {
-          // Issuer is not empty string when the module is imported by another module.
-          // Prevent from externalizing entry modules here.
-          if (data.contextInfo.issuer) {
-            // Node.js ECMAScript module loader does no extension searching.
-            // Add a file extension according to autoExtension config
-            // when data.request is a relative path and do not have an extension.
-            // If data.request already have an extension, we replace it with new extension
-            // This may result in a change in semantics,
-            // user should use copy to keep origin file or use another separate entry to deal this
-            let request: string = data.request;
-
-            const cssExternal = cssExternalHandler(
-              request,
-              callback,
-              jsExtension,
-              cssModulesAuto,
-              isStyleRedirect,
-            );
-
-            if (cssExternal !== false) {
-              return cssExternal;
+    resolvedJsRedirect: {
+      path: jsRedirectPath,
+      extension: jsRedirectExtension,
+    },
+    config: {
+      output: {
+        externals: [
+          async (data, callback) => {
+            const { request, getResolve, context, contextInfo } = data;
+            if (!request || !getResolve || !context || !contextInfo) {
+              return callback();
+            }
+
+            if (!resolver) {
+              resolver = (await getResolve()) as Resolver;
             }
 
-            if (request[0] === '.') {
-              const ext = extname(request);
+            // Issuer is not empty string when the module is imported by another module.
+            // Prevent from externalizing entry modules here.
+            if (contextInfo.issuer) {
+              let resolvedRequest: string = request;
+
+              const cssExternal = cssExternalHandler(
+                resolvedRequest,
+                callback,
+                jsExtension,
+                cssModulesAuto,
+                doesRedirectStyle,
+              );
+
+              if (cssExternal !== false) {
+                return cssExternal;
+              }
+
+              if (jsRedirectPath) {
+                try {
+                  resolvedRequest = await resolver(context, resolvedRequest);
+                } catch (e) {
+                  // Do nothing, fallthrough to other external matches.
+                  logger.debug(
+                    `Failed to resolve ${resolvedRequest} with resolver`,
```

Comment:
should the error be printed to debug info？

Reply:
I'm not sure which request failed to resolve. Bundleless is similar to transform, so regarding transform, we might be able to ignore the error for now.

Reply:
IMO, #251 means: the rewritten path should be "resovled", which is the issue addressed in this PR. The error to omit here is that a module can't be resolved. They're slightly different.

Reply:
@SoonIter What's your opinion?

Reply:
I think we can first come up with a reasonable use case for this error here. If we can't find one, we can assume that there is no use case, and then we can directly throw an error.

Reply:
Leave it here for now, we could change this in the future if needed.